rfc/rfc-0011-streaming-gateways.md

```diff
@@ -0,0 +1,68 @@
+# RFC-0011: Streaming Gateways
+
+**Authors:** @scothis, @markfisher, @ericbottard
+
+**Status:**
+
+**Pull Request URL:** [#1376](https://github.com/projectriff/riff/pull/1376)
+
+**Superseded by:** N/A
+
+**Supersedes:** N/A
+
+**Related:** N/A
+
+
+## Problem
+
+The relationships between a `Stream` and a `(InMemory|Kafka|Pulsar)Provider` is fragile. The stream contains a hostname for the provider's provisioner, which the stream controller calls when reconciling the stream. This has a few problems:
+- the `Stream` resource doesn't know when the `Provider` is unavailable, or deleted
+- the `Stream` has no way automatically obtain a new address for the provisioner
+- the provisioner name is completing in the `Service` namespace. The likelihood of a name collision is reduced by appending a suffix to the service name, however, the user must include this suffix when creating each stream
+- as new providers are created, there is a significant amount of reconciliation logic that is duplicated with trivial differences
+
+The problem is derived from [projectriff/system#86](https://github.com/projectriff/system/issues/86).
+
+### Anti-Goals
+
+This RFC will not prevent naming collisions between different provisioners with the same name. The first gateway of a given name wins.
+
+## Solution
+
+1. Rename `*Provider` to `*Gateway`
+
+   The high level role of the whole component is that of a gateway. 
+
+1. Introduce a `Gateway` resource which creates a single `Deployment` and `Service`
+
+   This reduces the toil of adding a new `*Gateway` reconciler as the boilerplate code will be encapsulated. It is recommend that users not create this resource directly.
+   
+   1. Cohabitate the gateway and provisioner `Deployment`s and `Service`s
+
+      A single `Deployment` with two containers and a single `Service` with two ports can replace the current two Deployments and two Services per provider.
+
+   1. Update each existing `*Gateway` to reconcile to a `Gateway`
```

Comment:
Maybe specify what a `Gateway` resource contains here then. I assume it's more than just the host as we discussed, if we want to do the heavy lifting in the generic `Gateway` reconciler: gateway (liiklus) image, provisioner image?  (both pod templates even I guess. Obviously the actual PR will have all the details, but I guess the RFC could benefit from some high level clarifications here)

Reply:
At a minimum I'd imaging we have a PodSpecTemplate that would become the core of the Deployment. I haven't worked through the details of the Gateway CRD shape.